### PR TITLE
ARROW-16329: [Java][C++] Keep more context when marshalling errors through JNI

### PR DESCRIPTION
**Pull request by zhztheplayer**
https://issues.apache.org/jira/browse/ARROW-16329

### CODE REVIEW

File: cpp/src/jni/dataset/jni_util.cc [hunk 1]
```diff
@@ -167,9 +167,16 @@ ReservationListenableMemoryPool::~ReservationListenableMemoryPool() {}
...
     env->ExceptionClear();
-    return Status::Invalid("Error during calling Java code from native code");
+    jclass describer_class =
+        env->FindClass("org/apache/arrow/dataset/jni/JniExceptionDescriber");
+    jmethodID describe_method = env->GetStaticMethodID(
+        describer_class, "describe", "(Ljava/lang/Throwable;)Ljava/lang/String;");
+    std::string description = JStringToCString(
+        env, (jstring)env->CallStaticObjectMethod(describer_class, describe_method, t));
+    return Status::Invalid("Error during calling Java code from native code: " +
+                           description);
```
Comment (pitrou) [inline cpp/src/jni/dataset/jni_util.cc:179]: Hmm... I don't know what is possible exactly, but ideally we should:
1) map Java exception classes to specific C++ status codes (for example a `IOException` would map to `Status::IOError` while `InvalidTypeException` would map to `Status::TypeError`)
2) keep the Java exception object in a Java-specific `StatusDetail` subclass, just like we do [for Python exceptions](https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/common.cc#L54-L139)

cc @lwhite1

Comment (pitrou) [reply cpp/src/jni/dataset/jni_util.cc:179]: Also this helper function should be moved out of dataset-specific code to be reused generically in all JNI bindings.

Comment (zhztheplayer) [reply cpp/src/jni/dataset/jni_util.cc:179]: > Also this helper function should be moved out of dataset-specific code to be reused generically in all JNI bindings.

Was considering doing the similar thing but maybe after we have a common code module for JNI stuffs?

Comment (zhztheplayer) [reply cpp/src/jni/dataset/jni_util.cc:179]: @pitrou sure, I have opened a ticket
https://issues.apache.org/jira/browse/ARROW-16691

Comment (pitrou) [reply cpp/src/jni/dataset/jni_util.cc:179]: @zhztheplayer Do you want to address the other parts of the initial comment in this PR?

Comment (zhztheplayer) [reply cpp/src/jni/dataset/jni_util.cc]: Addressed. Thanks for suggestion.

**Comment by zhztheplayer** [comment]
@pitrou @lidavidm I have updated the PR to address the early comments. Would you like to take a look again? Thanks.

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 2]
```diff
@@ -17,14 +17,14 @@
 
 #include <mutex>
 
+#include "./jni_util.h"
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_wrapper.cc]: done

File: java/dataset/src/main/cpp/jni_util.cc [hunk 3]
```diff
@@ -17,6 +17,7 @@
 
 #include "./jni_util.h"
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc]: done

File: java/dataset/src/main/cpp/jni_util.cc [hunk 4]
```diff
@@ -162,11 +165,104 @@ std::shared_ptr<ReservationListener> ReservationListenableMemoryPool::get_listen
...
+      describer_class, "describe", "(Ljava/lang/Throwable;)Ljava/lang/String;");
+  std::string description = JStringToCString(
+      env, (jstring)env->CallStaticObjectMethod(describer_class, describe_method, t));
+  return description;
+}
+
+arrow::Result<bool> IsErrorInstanceOf(JNIEnv* env, jthrowable t, std::string class_name) {
+  jclass jclass = env->FindClass(class_name.c_str());
+  if (jclass == nullptr) {
+    return arrow::Status::Invalid("Class not found: " + class_name);
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc]: Is `DCHECK` suitable for null-checking + returning? Anyway I have changed to return `false` in latest commit.

File: java/dataset/src/main/cpp/jni_util.cc [hunk 5]
```diff
@@ -162,11 +165,104 @@ std::shared_ptr<ReservationListener> ReservationListenableMemoryPool::get_listen
...
+}
+
+JNIEnv* GetEnvOrAttach(JavaVM* vm) {
+  JNIEnv* env;
+  int getEnvStat = vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION);
+  if (getEnvStat == JNI_EDETACHED) {
+    // Reattach current thread to JVM
+    getEnvStat = vm->AttachCurrentThread(reinterpret_cast<void**>(&env), nullptr);
+    if (getEnvStat != JNI_OK) {
+      std::cout << "Failed to attach current thread to JVM. " << std::endl;
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc]: done

File: java/dataset/src/main/java/org/apache/arrow/dataset/jni/JniExceptionDescriber.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.apache.arrow.dataset.jni;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+/**
+ * For native code to invoke to convert a java/lang/Throwable to jstring.
+ */
+public class JniExceptionDescriber {
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/java/org/apache/arrow/dataset/jni/JniExceptionDescriber.java]: Changed to use package private.

File: java/dataset/src/main/cpp/jni_util.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply java/dataset/src/main/cpp/jni_util.cc]: Well, DCHECK will abort the program (in a debug build). Presumably it's a bug if we can't find any of these classes/we shouldn't continue execution? But returning `false` is OK

File: java/dataset/src/main/cpp/jni_util.cc [hunk 7]
```diff
@@ -17,6 +17,7 @@
 
 #include "jni_util.h"
 
+#include <iostream>
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_util.cc:20]: Unused include now?

File: java/dataset/src/main/cpp/jni_util.cc [hunk 8]
```diff
@@ -162,11 +165,105 @@ std::shared_ptr<ReservationListener> ReservationListenableMemoryPool::get_listen
...
+  if (env == nullptr || this->cause_ == nullptr) {
+    return nullptr;
+  }
+  return (jthrowable)env->NewLocalRef(cause_);
+}
+
+std::string JavaErrorDetail::ToString() const {
+  JNIEnv* env = GetEnvOrAttach(vm_);
+  if (env == nullptr) {
+    return "Java Exception, ID: " + std::to_string(reinterpret_cast<size_t>(cause_));
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_util.cc:251]: cause_ is a pointer, so use uintptr_t?

File: java/dataset/src/main/cpp/jni_util.h [hunk 9]
```diff
@@ -91,6 +93,24 @@ void ReleaseNativeRef(jlong ref) {
   delete retrieved_ptr;
 }
 
+const char kJavaErrorDetailTypeId[] = "arrow::dataset::jni::JavaErrorDetail";
```
Comment (pitrou) [inline java/dataset/src/main/cpp/jni_util.h:96]: Should this be a `static` member inside `JavaErrorDetail`, or perhaps `extern` and defined in the `.cc`?
As is, it seems there will be symbol clashes if `jni_util.h` is included by several `.cc` files.

File: java/dataset/src/main/cpp/jni_util.cc [hunk 10]
```diff
@@ -162,11 +165,105 @@ std::shared_ptr<ReservationListener> ReservationListenableMemoryPool::get_listen
 
 ReservationListenableMemoryPool::~ReservationListenableMemoryPool() {}
 
+std::string Describe(JNIEnv* env, jthrowable t) {
+  jclass describer_class =
+      env->FindClass("org/apache/arrow/dataset/jni/JniExceptionDescriber");
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc:169]: done

File: java/dataset/src/main/cpp/jni_util.cc [hunk 11]
```diff
@@ -162,11 +165,105 @@ std::shared_ptr<ReservationListener> ReservationListenableMemoryPool::get_listen
...
+    ARROW_LOG(WARNING) << "Could not find Java class " << class_name;
+    return false;
+  }
+  return env->IsInstanceOf(t, jclass);
+}
+
+arrow::StatusCode MapJavaError(JNIEnv* env, jthrowable t) {
+  StatusCode code;
+  if (IsErrorInstanceOf(env, t, "org/apache/arrow/memory/OutOfMemoryException") ==
+      JNI_TRUE) {
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc]: done

File: java/dataset/src/main/cpp/jni_util.cc (hunk 8 above)
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.cc]: done

File: java/dataset/src/main/cpp/jni_util.h (hunk 9 above)
Comment (zhztheplayer) [reply java/dataset/src/main/cpp/jni_util.h]: I have moved the variable into `.cc`.

**Review by pitrou — APPROVED** [review]
I'll let a Java expert do the final approval but this looks basically ok to me (and a very marked improvement IMHO).

**Review by lidavidm — APPROVED** [review]

**Pull request merged**